File: test_lexer.py
import unittest
from tokenize_1 import tokenize, Token 

class TestLexer(unittest.TestCase):
    
    def test_palabras_reservadas(self):
        tokens = list(tokenize("const var print return break continue if else while func import true false"))
        palabras = ["CONST", "VAR", "PRINT", "RETURN", "BREAK", "CONTINUE", "IF", "ELSE", "WHILE", "FUNC", "IMPORT", "TRUE", "FALSE"]
        valores = ["const", "var", "print", "return", "break", "continue", "if", "else", "while", "func", "import", "true", "false"]
        for i, token in enumerate(tokens):
            self.assertEqual(token, Token(palabras[i], valores[i], 1))
    
    def test_identificadores(self):
        tokens = list(tokenize("variable _var1 a1b2c3"))
        self.assertEqual(tokens[0], Token("ID", "variable", 1))
        self.assertEqual(tokens[1], Token("ID", "_var1", 1))
        self.assertEqual(tokens[2], Token("ID", "a1b2c3", 1))
    
    def test_numeros(self):
        tokens = list(tokenize("123 45.67 .456 123."))
        self.assertEqual(tokens[0], Token("INTEGER", "123", 1))
        self.assertEqual(tokens[1], Token("FLOAT", "45.67", 1))
        self.assertEqual(tokens[2], Token("FLOAT", ".456", 1))
        self.assertEqual(tokens[3], Token("FLOAT", "123.", 1))
    
    def test_operadores(self):
        tokens = list(tokenize("+ - * / <= >= == != && || ^"))
        tipos = ["PLUS", "MINUS", "TIMES", "DIVIDE", "LE", "GE", "EQ", "NE", "LAND", "LOR", "GROW"]
        valores = ["+", "-", "*", "/", "<=", ">=", "==", "!=", "&&", "||", "^"]
        for i, token in enumerate(tokens):
            self.assertEqual(token, Token(tipos[i], valores[i], 1))
    
    def test_simbolos_miscelaneos(self):
        tokens = list(tokenize("= ; ( ) { } , `"))
        tipos = ["ASSIGN", "SEMI", "LPAREN", "RPAREN", "LBRACE", "RBRACE", "COMMA", "DEREF"]
        valores = ["=", ";", "(", ")", "{", "}", ",", "`"]
        for i, token in enumerate(tokens):
            self.assertEqual(token, Token(tipos[i], valores[i], 1))
    
    def test_comentarios(self):
        tokens = list(tokenize("var x = 5; // comentario"))
        self.assertEqual(len(tokens), 5)
        self.assertEqual(tokens[0], Token("VAR", "var", 1))
        self.assertEqual(tokens[1], Token("ID", "x", 1))
        self.assertEqual(tokens[2], Token("ASSIGN", "=", 1))
        self.assertEqual(tokens[3], Token("INTEGER", "5", 1))
        self.assertEqual(tokens[4], Token("SEMI", ";", 1))
    
    def test_comentario_bloque(self):
        tokens = list(tokenize("var x = 5; /* comentario */ var y = 10;"))
        self.assertEqual(len(tokens), 10)
    
    def test_error_caracter(self):
        with self.assertRaises(ValueError):
            list(tokenize("@"))
    
    def test_error_comilla_no_cerrada(self):
        with self.assertRaises(ValueError):
            list(tokenize("'a"))
    
    def test_error_comentario_no_cerrado(self):
        with self.assertRaises(ValueError):
            list(tokenize("/* comentario sin cerrar"))
    
if __name__ == '__main__':
    unittest.main()

File: tests/test_lexer.rs
#[cfg(test)]
mod test_lexer {
    use gox_compiler::lexer::{
        ControlKeyword, DeclarationKeyword, LiteralType, OneCharToken, TokenType, TwoCharToken,
        tokenize,
    };

    #[test]
    fn test_keywords() {
        let input = "if while const var";
        let tokens = tokenize(input);

        assert_eq!(tokens.len(), 5);
        assert_eq!(
            tokens[0].token_type,
            TokenType::ControlKeyword(ControlKeyword::If)
        );
        assert_eq!(
            tokens[1].token_type,
            TokenType::ControlKeyword(ControlKeyword::While)
        );
        assert_eq!(
            tokens[2].token_type,
            TokenType::DeclarationKeyword(DeclarationKeyword::Const)
        );
        assert_eq!(
            tokens[3].token_type,
            TokenType::DeclarationKeyword(DeclarationKeyword::Var)
        );
    }

    #[test]
    fn test_literals() {
        let input = "123 3.14 true false 'a'";
        let tokens = tokenize(input);

        assert_eq!(tokens.len(), 6);
        assert_eq!(
            tokens[0].token_type,
            TokenType::Literal(LiteralType::Integer)
        );
        assert_eq!(tokens[1].token_type, TokenType::Literal(LiteralType::Float));
        assert_eq!(tokens[2].token_type, TokenType::Literal(LiteralType::Bool));
        assert_eq!(tokens[3].token_type, TokenType::Literal(LiteralType::Bool));
        assert_eq!(tokens[4].token_type, TokenType::Literal(LiteralType::Char));
    }

    #[test]
    fn test_identifiers() {
        let input = "var x my_var _privateVar";
        let tokens = tokenize(input);

        assert_eq!(tokens.len(), 5);
        assert_eq!(
            tokens[0].token_type,
            TokenType::DeclarationKeyword(DeclarationKeyword::Var)
        );
        assert_eq!(tokens[1].token_type, TokenType::Identifier);
        assert_eq!(tokens[2].token_type, TokenType::Identifier);
        assert_eq!(tokens[3].token_type, TokenType::Identifier);
    }

    #[test]
    fn test_one_char_operators() {
        let input = "+ - * / < > ^ = ; ( ) { } , ` !";
        let tokens = tokenize(input);

        assert_eq!(tokens.len(), 17);
        assert_eq!(tokens[0].token_type, TokenType::OneChar(OneCharToken::Plus));
        assert_eq!(
            tokens[1].token_type,
            TokenType::OneChar(OneCharToken::Minus)
        );
        assert_eq!(
            tokens[2].token_type,
            TokenType::OneChar(OneCharToken::Times)
        );
        assert_eq!(
            tokens[3].token_type,
            TokenType::OneChar(OneCharToken::Divide)
        );
        assert_eq!(
            tokens[4].token_type,
            TokenType::OneChar(OneCharToken::Lt)
        );
        assert_eq!(
            tokens[5].token_type,
            TokenType::OneChar(OneCharToken::Gt)
        );

        assert_eq!(
            tokens[6].token_type,
            TokenType::OneChar(OneCharToken::Grow)
        );
        assert_eq!(
            tokens[7].token_type,
            TokenType::OneChar(OneCharToken::Assign)
        );
        assert_eq!(
            tokens[8].token_type,
            TokenType::OneChar(OneCharToken::Semi)
        );
        assert_eq!(
            tokens[9].token_type,
            TokenType::OneChar(OneCharToken::LParen)
        );
        assert_eq!(
            tokens[10].token_type,
            TokenType::OneChar(OneCharToken::RParen)
        );
        assert_eq!(
            tokens[11].token_type,
            TokenType::OneChar(OneCharToken::LBrace)
        );
        assert_eq!(
            tokens[12].token_type,
            TokenType::OneChar(OneCharToken::RBrace)
        );
        assert_eq!(
            tokens[13].token_type,
            TokenType::OneChar(OneCharToken::Comma)
        );
        assert_eq!(
            tokens[14].token_type,
            TokenType::OneChar(OneCharToken::Deref)
        );
        assert_eq!(
            tokens[15].token_type,
            TokenType::OneChar(OneCharToken::Not)
        );
    }

    #[test]
    fn test_two_char_operators() {
        let input = "&& || <= >= == !=";
        let tokens = tokenize(input);

        assert_eq!(tokens.len(), 7);
        assert_eq!(tokens[0].token_type, TokenType::TwoChar(TwoCharToken::Land));
        assert_eq!(tokens[1].token_type, TokenType::TwoChar(TwoCharToken::Lor));
        assert_eq!(tokens[2].token_type, TokenType::TwoChar(TwoCharToken::Le));
        assert_eq!(tokens[3].token_type, TokenType::TwoChar(TwoCharToken::Ge));
        assert_eq!(tokens[4].token_type, TokenType::TwoChar(TwoCharToken::Eq));
        assert_eq!(tokens[5].token_type, TokenType::TwoChar(TwoCharToken::Ne));
    }

    #[test]
    fn test_whitespace_and_newlines() {
        let input = "if   x > 10 \n while x < 5";
        let tokens = tokenize(input);

        assert_eq!(tokens.len(), 9);
        assert_eq!(
            tokens[0].token_type,
            TokenType::ControlKeyword(ControlKeyword::If)
        );
        assert_eq!(tokens[1].token_type, TokenType::Identifier);
        assert_eq!(tokens[2].token_type, TokenType::OneChar(OneCharToken::Gt));
        assert_eq!(
            tokens[3].token_type,
            TokenType::Literal(LiteralType::Integer)
        );
        assert_eq!(
            tokens[4].token_type,
            TokenType::ControlKeyword(ControlKeyword::While)
        );
        assert_eq!(tokens[5].token_type, TokenType::Identifier);
    }

    #[test]
    fn test_unknown_characters() {
        let input = "@ # $ % ^ as.a &";
        let tokens = tokenize(input);

        assert_eq!(tokens.len(), 8);
        assert_eq!(tokens[0].token_type, TokenType::Illegal);
        assert_eq!(tokens[1].token_type, TokenType::Illegal);
        assert_eq!(tokens[2].token_type, TokenType::Illegal);
        assert_eq!(tokens[3].token_type, TokenType::Illegal);
        assert_eq!(tokens[4].token_type, TokenType::OneChar(OneCharToken::Grow));
        assert_eq!(tokens[5].token_type, TokenType::Illegal);
        assert_eq!(tokens[6].token_type, TokenType::Illegal);
    }

    #[test]
    fn test_eof_token() {
        let input = "var x = 10";
        let tokens = tokenize(input);

        assert_eq!(tokens[tokens.len() - 1].token_type, TokenType::EOF);
    }

    #[test]
    fn test_comments() {
        let input = "// This is a comment\n var x = 10 // another comment";
        let tokens = tokenize(input);

        assert_eq!(tokens.len(), 5); // El número de tokens no debe contar los comentarios
        assert_eq!(
            tokens[0].token_type,
            TokenType::DeclarationKeyword(DeclarationKeyword::Var)
        );
        assert_eq!(tokens[1].token_type, TokenType::Identifier);
        assert_eq!(
            tokens[2].token_type,
            TokenType::OneChar(OneCharToken::Assign)
        );
        assert_eq!(
            tokens[3].token_type,
            TokenType::Literal(LiteralType::Integer)
        );
    }
}
